src/overlays/actors/ovl_En_Fr/z_en_fr.c

```diff
@@ -27,99 +183,1134 @@ const ActorInit En_Fr_InitVars = {
+        this->growingScaleIndex = 0;
+        this->isActive = false;
+        this->isJumpingToFrogSong = false;
+        this->songIndex = 7;
+        this->unusedButterflyActor = NULL;
+
+        EnFr_OrientUnderwater(this);
+        EnFr_DrawIdle(this);
+
+        this->actor.update = EnFr_UpdateActive;
+        this->isButterflyDrawn = false;
+        this->xyAngleButterfly = 0x1000 * (s16) Math_Rand_ZeroFloat(255.0f);
+        this->posButterflyLight.x = this->posButterfly.x = this->posLogSpot.x;
+        this->posButterflyLight.y = this->posButterfly.y = this->posLogSpot.y + 50.0f;
+        this->posButterflyLight.z = this->posButterfly.z = this->posLogSpot.z;
+
+        this->actor.flags &= ~1;
+
+    }
+}
+#else
 #pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/EnFr_Update.s")
+#endif
+
+
+void EnFr_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    EnFr* this = THIS;
+    LightContext_RemoveLight(globalCtx, &globalCtx->lightCtx, this->lightNode);
+}
+
+void EnFr_IsDivingIntoWater(EnFr* this, GlobalContext* globalCtx) {
+    WaterBox* waterBox;
+    f32 waterSurface;
+
+    if (func_80042244(globalCtx, &globalCtx->colCtx, this->actor.posRot.pos.x, this->actor.posRot.pos.z, &waterSurface,
+                      &waterBox)) {
+        this->isBelowWaterSurfacePrevious = this->isBelowWaterSurfaceCurrent;
+        if (this->actor.posRot.pos.y <= waterSurface) {
+            this->isBelowWaterSurfaceCurrent = true;
+        } else {
+            this->isBelowWaterSurfaceCurrent = false;
+        }
+    }
+}
+
+void EnFr_DivingIntoWater(EnFr* this, GlobalContext* globalCtx) {
+    Vec3f vec;
+
+    // Jumping into or out of water
+    if (this->isBelowWaterSurfaceCurrent != this->isBelowWaterSurfacePrevious) {
+        vec.x = this->actor.posRot.pos.x;
+        vec.y = this->actor.posRot.pos.y - 10.0f;
+        vec.z = this->actor.posRot.pos.z;
+        EffectSsGSplash_Spawn(globalCtx, &vec, NULL, NULL, 1, 1);
+
+        if (this->isBelowWaterSurfaceCurrent == false) {
+            // Jumping out of the water
+            Audio_PlayActorSound2(&this->actor, NA_SE_EV_DIVE_INTO_WATER_L);
+        } else { // Jumping into the water
+            Audio_PlayActorSound2(&this->actor, NA_SE_EV_BOMB_DROP_WATER);
+        }
+    }
+}
+
+// Check if frog is moving down and is below posLogSpot. If so, then set frog to posLogSpot.
+s32 EnFr_IsBelowLogSpot(EnFr* this, f32* yDistToLogSpot) {
+    yDistToLogSpot[0] = this->actor.posRot.pos.y - this->posLogSpot.y;
+
+    if ((this->actor.velocity.y < 0.0f) && (yDistToLogSpot[0] <= 0.0f)) {
+        this->actor.velocity.y = 0.0f;
+        this->actor.posRot.pos.y = this->posLogSpot.y;
+        return true;
+    } else {
+        return false;
+    }
+}
+
+s32 EnFr_IsAboveAndWithin30xzDist(Player* player, EnFr* this) {
+    f32 xDistFromLink;
+    f32 zDistFromLink;
+    f32 yDistFromLink;
+
+    xDistFromLink = player->actor.posRot.pos.x - this->actor.posRot.pos.x;
```

Comment:
you can do the assignment on the same line as the declare if you want, will make this more compact